packages/peregrine/src/Toasts/useToastContext.js

```diff
@@ -0,0 +1,127 @@
+import React, { createContext, useContext, useReducer } from 'react';
+/**
+ * The state of the toast store.
+ * @typedef {Object} ToastState
+ * @property {Object} toasts - unordered map of id:toast
+ */
+const initialState = {
+    toasts: {}
+};
+
+const reducer = (prevState = initialState, action = {}) => {
+    const { type, payload } = action;
+
+    if (!type) {
+        throw new TypeError('action.type is required');
+    }
+
+    if (!payload.id) {
+        throw new TypeError('action.payload.id is required');
+    }
+
+    switch (type) {
+        case 'add': {
+            // If we are adding a toast that already exists we need to clear the
+            // old removal timeout effectively resetting the delete timer.
+            if (prevState.toasts[payload.id]) {
+                window.clearTimeout(
+                    prevState.toasts[payload.id].removalTimeoutId
+                );
+            }
+
+            const duplicate = !!prevState.toasts[payload.id];
+
+            // For duplicate toasts, do not update the timestamp to maintain
+            // order of toast emission.
+            const timestamp = duplicate
+                ? prevState.toasts[payload.id].timestamp
+                : Date.now();
+
+            // Use a random key to trigger a recreation of this component if it
+            // is a duplicate so that we can re-trigger the blink animation.
+            const key = duplicate ? Math.random() : payload.id;
+
+            const newState = {
+                ...prevState,
+                toasts: {
+                    ...prevState.toasts,
+                    [payload.id]: {
+                        ...payload,
+                        timestamp,
+                        duplicate,
+                        key
+                    }
+                }
+            };
+
+            return newState;
+        }
+        case 'remove': {
+            const newState = {
+                ...prevState
+            };
+
+            // It is possible to attempt to delete a non-existant toast so let's
+            // gracefully handle that.
+            if (newState.toasts[payload.id]) {
+                // Clear the old toast removal timeout incase an identical toast
+                // is added after this.
+                window.clearTimeout(
+                    newState.toasts[payload.id].removalTimeoutId
+                );
```

Comment:
That would require access to the `removalTimeoutId` from the toast in the store. We don't have access to that in the `removeToast` function.

Reply:
Perhaps we should, though. If we follow the pattern from `useQueryResult`, the following is possible:

```es6
export const useToasts = () => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const { toasts } = state

  const addToast = useCallback(/* ... */)

  const removeToast = useCallback(
    () => {
      payload => {
        const timeoutId = toasts.get(payload)

        // `timeoutId` doesn't have to match a timeout
        // in fact, it can even be `null` or `undefined`
        window.clearTimeout(timeoutId)

        dispatch({
          payload,
          type: "remove"
        })
      }
    },
    [toasts]
  )

  const api = useMemo(
    () => ({
      addToast,
      dispatch,
      removeToast
    }),
    []
  )

  return [state, api]
}
```

Reply:
Got it. I'll give this a shot.